File: sample_4/ANSWER/verlet.cc
#define xm 10.0
#include <cmath>
#include "md.h"

double dnint(double x);

void verlet_1(particle *q, double dt)
{
  int i, j;
  double x;
  for (i=0;i<N;i++) {
    for (j=0;j<3;j++) {      
      q[i].x[j] += dt * q[i].v[j] + 0.5*dt*dt*q[i].f[j]/xm;
      q[i].x[j] = q[i].x[j] - l_box*dnint(q[i].x[j]/l_box);
      q[i].v[j] += 0.5*dt*q[i].f[j]/xm;
    }
  }
}

void verlet_2(particle *q, double dt, double &e_kinetic)
{
  int i, j;

  e_kinetic = 0.0;
  for (i=0;i<N;i++) {
    for (j=0;j<3;j++) {
      q[i].v[j] += 0.5*dt*q[i].f[j]/xm;
      e_kinetic += 0.5*xm*q[i].v[j]*q[i].v[j];
    }
  }
}

File: MD_sample/OpenACC_F90/verlet.f90
subroutine verlet_1(dt)
!$acc routine
  use md
  implicit none
  integer(gi):: i, k
  real   (gp):: dt
  !$acc loop gang private(i,k)
  do i=1, N3
     k = i/3 + 1
     xx(i) = xx(i) + dt*vv(i) + 0.5_gp*dt*dt*ff(i)/xm(k)
     xx(i) = xx(i) - l_box*dnint(xx(i)/l_box)
     vv(i) = vv(i) + 0.5_gp*dt*ff(i)/xm(k)
  end do
  return
end subroutine verlet_1
!
!################################################################
subroutine verlet_2(dt)
!$acc routine
  use md
  implicit none
  integer(gi):: i, k
  real   (gp):: dt
  !$acc loop gang private(i,k)
  do i=1, N3
     k = i/3 + 1
     vv(i) = vv(i) + 0.5_gp*dt*ff(i)/xm(k)
  end do
  return
end subroutine verlet_2

